components/Init/src/virtio_vsock.c

```diff
@@ -0,0 +1,183 @@
+static int virtio_vsock_notify_recv(virtqueue_device_t *queue)
+{
+    int err;
+    void *buf = NULL;
+    size_t buf_size = 0;
+    vq_flags_t flag;
+    virtqueue_ring_object_t handle;
+
+    while (virtqueue_get_available_buf(queue, &handle)) {
+        char temp_buf[VIRTIO_VSOCK_CAMKES_MTU];
+        size_t len = virtqueue_scattered_available_size(queue, &handle);
+        if (camkes_virtqueue_device_gather_copy_buffer(queue, &handle, (void *)temp_buf, len) < 0) {
+            ZF_LOGW("Dropping data: Can't gather vq buffer.");
+            continue;
+        }
+
+        vsock_rx_complete(virtio_vsock->emul, temp_buf, len);
+        queue->notify();
+    }
+}
+
+void handle_vsock_irq(vm_t *vmm)
+{
+    /* Poll each ring buffer to see if data was added */
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].recv_queue && VQ_DEV_POLL(connections[i].recv_queue)) {
+            virtio_vsock_notify_recv(connections[i].recv_queue);
+        }
+        if (connections[i].send_queue && VQ_DRV_POLL(connections[i].send_queue)) {
+            virtio_vsock_notify_free_send(connections[i].send_queue);
+        }
+    }
+}
+
+static void tx_forward(int cid, void *buffer, unsigned int len)
+{
+    /* Find the CID */
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].cid == cid) {
+            int err = camkes_virtqueue_driver_scatter_send_buffer(connections[i].send_queue, buffer, len);
+            connections[i].send_queue->notify();
+            return;
+        }
+    }
+
+    ZF_LOGE("Invalid CID not found: %d", cid);
+}
+
+static void make_virtio_vsock_virtqueues(void)
+{
+    int err;
+    for (int i = 0; i < NUM_PORTS; i++) {
+        virtqueue_driver_t *vq_send;
+        virtqueue_device_t *vq_recv;
+
+        vq_recv = malloc(sizeof(*vq_recv));
+        ZF_LOGF_IF(!vq_recv, "Unable to alloc recv camkes-virtqueue for port: %d", i);
+
+        vq_send = malloc(sizeof(*vq_send));
+        ZF_LOGF_IF(!vq_send, "Unable to alloc send camkes-virtqueue for port: %d", i);
+
+        /* Initialise send virtqueue */
+        err = camkes_virtqueue_driver_init(vq_send, socket_layout[i].send_id);
+        ZF_LOGF_IF(err, "Unable to initialise send camkes-virtqueue for port: %d", i);
+
+        /* Initialise recv virtqueue */
+        err = camkes_virtqueue_device_init(vq_recv, socket_layout[i].recv_id);
+        ZF_LOGF_IF(err, "Unable to initialise recv camkes-virtqueue for port: %d", i);
+
+
+        connections[i].cid = socket_layout[i].cid;
+        connections[i].recv_queue = vq_recv;
+        connections[i].send_queue = vq_send;
+    }
+}
+
+static void vsock_inject_irq(void *cookie)
+{
+    virtio_vsock_cookie_t *virtio_cookie = (virtio_vsock_cookie_t *)cookie;
+    if (!virtio_cookie) {
+        ZF_LOGE("NULL virtio cookie given to raw irq handler");
+        return;
+    }
+
+    int err = vm_inject_irq(virtio_cookie->vm->vcpus[BOOT_VCPU], VIRTIO_SCK_IRQ);
```

Comment:
We *always* inject into the BOOT VCPU, instead of allowing the guest to choose a CPU to be interrupted via emulated the IOAPIC/GIC ?  Why?

Reply:
I'm not sure if the current VMM code handles IRQ affinity changing for the guest, if it does, can anyone point me to the right place in the VMM code?